### ISSUE TITLE
[Feature Request] Property to set prestart task to run only one instance even if in larger group

### ISSUE THREAD

**Issue by tommyalatalo** [hashicorp/nomad#8067]
It would be great to be able to set a property for a prestart task to run only as a single instance even if the task group has a count larger than 1.

The use case for this is that we are using prestart tasks to populate Vault secrets, and if the group has a count larger than 1 then the prestart task will also be executed multiple times, resulting in multiple versions of the same secrets being updated in the vault secret kv in our case.

So something like the below is what I'm proposing, I redacted all other job stanzas to keep the example clear. The lifecycle "mode" property could consist of options "single" and "group" to begin with, defaulting to "group", meaning it will start up as many instances as the group count declares.

```
group "example" {
    count = 3

    task "prestart" {
      lifecycle {
        hook = "prestart"
        mode = "single"
      }

      driver = "docker"

      config {
        image              = "alpine:3.11"
        entrypoint         = ["myPrestart.sh"]
      }
    }

    task "main" {
      driver = "docker"

      config {
        image              = "myImage"
      }
    }
```

**Comment by tgross** [comment]
Hi @tommyalatalo!

The way I'd generalize this feature request is to say that you're looking for a way to have _task group_ dependencies, and not just task dependencies. Nomad places whole task groups together as an allocation, so splitting out your prestart task gets you really close but doesn't have the lifecycle hook.

As a workaround in the meantime, I'd solve this problem by having the prestart task for every task group, but have the prestart tasks obtain a lock in Consul to assert that only one of them actually executes. Or if you don't need that level of guarantee, use the `NOMAD_ALLOC_INDEX` to make sure that only the prestart task with `NOMAD_ALLOC_INDEX = 0` updates the secret.

**Comment by tommyalatalo** [comment]
> Hi @tommyalatalo!
> 
> The way I'd generalize this feature request is to say that you're looking for a way to have _task group_ dependencies, and not just task dependencies. Nomad places whole task groups together as an allocation, so splitting out your prestart task gets you really close but doesn't have the lifecycle hook.
> 
> As a workaround in the meantime, I'd solve this problem by having the prestart task for every task group, but have the prestart tasks obtain a lock in Consul to assert that only one of them actually executes. Or if you don't need that level of guarantee, use the `NOMAD_ALLOC_INDEX` to make sure that only the prestart task with `NOMAD_ALLOC_INDEX = 0` updates the secret.

Yes, you're absolutely right in that a "group prestart task" is synonymous with what I'm requesting.

How would I make sure that the prestart task with NOMAD_ALLOC_INDEX = 0 is the only one to update the secret, I'm not sure how you mean?

**Comment by tommyalatalo** [comment]
> Hi @tommyalatalo!
> 
> The way I'd generalize this feature request is to say that you're looking for a way to have _task group_ dependencies, and not just task dependencies. Nomad places whole task groups together as an allocation, so splitting out your prestart task gets you really close but doesn't have the lifecycle hook.
> 
> As a workaround in the meantime, I'd solve this problem by having the prestart task for every task group, but have the prestart tasks obtain a lock in Consul to assert that only one of them actually executes. Or if you don't need that level of guarantee, use the `NOMAD_ALLOC_INDEX` to make sure that only the prestart task with `NOMAD_ALLOC_INDEX = 0` updates the secret.

I actually tried using NOMAD_ALLOC_INDEX for this, it works with "service" type jobs, but with system jobs all of the prestart tasks get an NOMAD_ALLOC_INDEX of 0. So this workaround is actually not feasible.

**Comment by urusha** [comment]
While we have a workaround for services, `lifecycle` option would be much better, and the same for `poststart` tasks.

**Comment by zhujinhe** [comment]
Really looking forward this feature.  In my scenario, when deploying multiple Web instances, Django for example, we only need to migrate the database once when we update the docker image.

**Comment by nocive** [comment]
A much needed feature for us too.

**Comment by dmoss18** [comment]
I was trying to use lifecycle hooks for running database migrations automatically. It would be very useful if I could specify a "group-level" lifecycle task

Or is there a way to have a job that triggers another job? i.e. I have a database migration job, and it starts my web server job when finished?
